testbed/tests/window/test_window.py

```diff
@@ -521,47 +602,326 @@ async def test_move_and_resize(second_window, second_window_probe):
+            (WindowState.PRESENTATION, WindowState.FULLSCREEN),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "second_window_class, second_window_kwargs",
+        [
+            (
+                toga.MainWindow,
+                dict(title="Secondary Window", position=(200, 150)),
+            )
+        ],
+    )
+    async def test_window_state_direct_change(
+        app,
+        app_probe,
+        initial_state,
+        final_state,
+        second_window,
+        second_window_probe,
+        intermediate_states=tuple(
+            random.sample(
+                [
+                    WindowState.NORMAL,
+                    WindowState.MINIMIZED,
+                    WindowState.MAXIMIZED,
+                    WindowState.FULLSCREEN,
+                    WindowState.PRESENTATION,
+                ],
+                5,
+            )
+        ),
+    ):
+        "Window state can be directly changed to another state."
+        if (
+            WindowState.MINIMIZED in {initial_state, final_state}
+            and not second_window_probe.supports_minimize
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+        elif (
+            WindowState.FULLSCREEN in {initial_state, final_state}
+            and not second_window_probe.supports_fullscreen
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+        elif (
+            WindowState.PRESENTATION in {initial_state, final_state}
+            and not second_window_probe.supports_presentation
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+
+        second_window.toolbar.add(app.cmd1)
+        second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
+        second_window.show()
+
+        # Add delay to ensure windows are visible after animation.
+        await second_window_probe.wait_for_window(
+            "Secondary window is visible", full_screen=True
+        )
+
+        assert second_window_probe.get_window_state() == WindowState.NORMAL
 
+        # Set to initial state
+        second_window.state = initial_state
+        # Add delay to ensure windows are visible after animation.
+        await second_window_probe.wait_for_window(
+            f"Secondary window is in {initial_state}", full_screen=True
+        )
+
+        assert second_window_probe.get_window_state() == initial_state
+
+        # Set to the intermediate states but don't wait for the OS delay.
+        for state in intermediate_states:
+            second_window.state = state
+
+        # Set to final state
+        second_window.state = final_state
+        # Add delay to ensure windows are visible after animation.
+        await app_probe.redraw(f"Secondary window is in {final_state}", delay=1.5)
```

Comment:
This sort of delay is what `wait_for_window` is for - some platforms can move faster than others.